static/js/usernameTrie.js

```diff
@@ -4,23 +4,29 @@
 // Initialized empty: `const trie = UsernameTrie()`
 // Returns an object that supports the following methods:
 //  `insertUser({ username: string, firstName: string, lastName: string })` - inserts the user's names into the trie
-//  `findAllStartingWith(prefix)` - returns an object mapping username to `{ firstName, lastName }` for all user
-//    objects in the trie with at least one of username, first name or last name matching the given prefix
+//  `findAllStartingWith(prefix, limit = Infinity)` - returns an object mapping username to `{ firstName, lastName }` for all user
+//    objects in the trie with at least one of username, first name or last name matching the given prefix,
+//    up to the given limit if provided, otherwise all matches are returned.
 
 const UsernameTrie = () => {
   const trieRoot = {};
 
-  const getAllMatchesFromHere = (trieNode) => {
+  const getAllMatchesFromHere = (trieNode, limit) => {
     const matches = {};
+    let numMatches = 0;
 
     const traverseAndSaveMatches = (currNode = trieNode) => {
       if (currNode.matchTerminatesHere) {
         Object.keys(currNode.matchingUsersMap).forEach((username) => {
-          if (!matches[username]) {
+          if (!matches[username] && numMatches < limit) {
```

Comment:
Why is the `numMatches < limit` check needed here?

Reply:
In case there is more than one match terminating at the current node.  I suppose in our case that's not possible because each username is unique, but this doesn't hurt to have in here and keeps the implementation more general/extensible

Reply:
Ah, no, there can indeed be multiple matches ending at a given node, because we also store first names and last names in the trie. So this is needed to avoid going over the limit